kernel/kernel-api/src/main/java/io/delta/kernel/internal/tablefeatures/TableFeature.java

```diff
@@ -0,0 +1,142 @@
+ * feature-specific table properties are set. For example, `changeDataFeed` is automatically
+ * supported when there's a table property `delta.enableChangeDataFeed=true`. See {@link
+ * FeatureAutoEnabledByMetadata} for details on how to define such features. This is independent of
+ * the table's enabled features. When a feature is supported (explicitly or implicitly) by the table
+ * protocol but its metadata requirements are not satisfied, then clients still have to understand
+ * the feature (at least to the extent that they can read and preserve the existing data in the
+ * table that uses the feature).
+ */
+public abstract class TableFeature {
+  private final String featureName;
+  private final int minReaderVersion;
+  private final int minWriterVersion;
+
+  /**
+   * Constructor. Does validations to make sure:
+   *
+   * <ul>
+   *   <li>Feature name is not null or empty and has valid characters
+   *   <li>minReaderVersion is always 0 for writer features
+   *   <li>all legacy features can be auto applied based on the metadata and protocol
+   *
+   * @param featureName a globally-unique string indicator to represent the feature. All characters
+   *     must be letters (a-z, A-Z), digits (0-9), '-', or '_'. Words must be in camelCase.
+   * @param minReaderVersion the minimum reader version this feature requires. For a feature that
+   *     can only be explicitly supported, this is either `0` or `3` (the reader protocol version
+   *     that supports table features), depending on the feature is writer-only or reader-writer.
+   *     For a legacy feature that can be implicitly supported, this is the first protocol version
+   *     which the feature is introduced.
+   * @param minWriterVersion the minimum writer version this feature requires. For a feature that
+   *     can only be explicitly supported, this is the writer protocol `7` that supports table
+   *     features. For a legacy feature that can be implicitly supported, this is the first protocol
+   *     version which the feature is introduced.
+   */
+  public TableFeature(String featureName, int minReaderVersion, int minWriterVersion) {
+    this.featureName = requireNonNull(featureName, "name is null");
+    checkArgument(!featureName.isEmpty(), "name is empty");
+    checkArgument(
+        featureName.chars().allMatch(c -> Character.isLetterOrDigit(c) || c == '-' || c == '_'),
+        "name contains invalid characters: " + featureName);
+    this.minReaderVersion = minReaderVersion;
+    this.minWriterVersion = minWriterVersion;
+
+    validate();
+  }
+
+  /** @return the name of the table feature. */
+  public String featureName() {
+    return featureName;
+  }
+
+  /**
+   * @return true if this feature is applicable to both reader and writer, false if it is
+   *     writer-only.
+   */
+  public boolean isReaderWriterFeature() {
+    return this instanceof TableFeatures.ReaderWriterFeatureType;
+  }
```

Comment:
Maybe a silly question, but do we need this interface at all? Can we just infer this from `minReaderVersion`?

Reply:
I think in your doc you mentioned this might do a check and require that `minReaderVersion` is overriden to be non-zero?

Reply:
Hmm I guess I see it is used below in `validate`. Still wondering what value the interface uses though and if we would just always use this method rather than checking instanceof

Reply:
It is a descriptive way. When you see a table feature and what is inheirting, we know it is a readerwriter or writer only feature.

we could depend on minReaderVersion = 0, but that's just hacky.